clash-lib/src/Clash/Netlist/BlackBox/Util.hs

```diff
@@ -307,22 +315,84 @@ setSym bbCtx l = do
                   ++ "blackbox for " ++ bbnm
         )
 
-selectNewName
-    :: Foldable t
-    => t String
-    -- ^ Set of existing names
-    -> FilePath
-    -- ^ Name for new file (
-    -> String
+type FileName = FilePath
+
+-- | Canonicalizes a datafile path, keeping in mind that the path might be
+-- relative to one of the import directories passed in with @-i@.
+canonicalizeDataFilePath
+  :: [FilePath]
+  -- ^ Import directories passed in with @-i@
+  -> FilePath
+  -- ^ File to canonicalize
+  -> IO FilePath
+canonicalizeDataFilePath idirs toCanonicalize
+  | isAbsolute toCanonicalize =
+    ifM
+      (doesFileExist toCanonicalize)
+      (canonicalizePath toCanonicalize)
+      (error [I.i|Could not find data file #{show toCanonicalize}. Does it exist?|])
+  | otherwise = do
+      let candidates = map (</> toCanonicalize) idirs
+      found <- filterM doesFileExist candidates
+      case found of
+        [] -> error [I.i|
+          Could not find data file #{show toCanonicalize}. The following directories were
+          searched:
+            #{idirs}
+        |]
+        (_:_:_) -> error [I.i|
+          Multiple data files for #{show toCanonicalize} found. The following candidates
+          were found:
+            #{found}
+          Please disambiguate data files.
+        |]
+        [c] ->
+          canonicalizePath c
+
+
+-- | Select a new file name that doesn't collide with existing names. Might return
+-- the same name as given if there are no collisions.
+selectNewName ::
+  -- | Set of existing names
+  Set FileName ->
+  -- | Name for file
+  FileName ->
+  -- | Updated file name
+  FileName
 selectNewName as a
-  | elem a as = selectNewName as (replaceBaseName a (takeBaseName a ++ "_"))
+  | a `Set.member` as = selectNewName as (replaceBaseName a (takeBaseName a ++ "_"))
   | otherwise = a
 
-renderFilePath :: [(String,FilePath)] -> String -> ([(String,FilePath)],String)
-renderFilePath fs f = ((f'',f):fs, f'')
-  where
-    f'  = takeFileName f
-    f'' = selectNewName (map fst fs) f'
+-- | Add a data file to a collection of existing files. Typically used in the HDL
+-- backend's 'addAndSetData'.
+addFilePath ::
```

Comment:
Oh don't forget to add `NONLINE` as @rowanG077 mentions in the issue thread.

Reply:
I opted to make `canonicalizeDataFilePath` pure and add NOINLINE there :)